### ISSUE TITLE
Cannot restore restic backup

### ISSUE THREAD

**Issue by domcar** [heptio/ark#615]
Hi,
i'm testing the 0.9.0_alpha_2 version of ark. I deployed ark (and restic) on a local cluster and I'm using Minio for the bucket (two different bucktes, one for ark and one for restic).
I was able to execute a backup and I can see that files have been backed up. But the restore process didn't work out:
```
./ark restore describe restic-20180627115342
Name:         restic-20180627115342
Namespace:    heptio-ark
Labels:       <none>
Annotations:  <none>
Backup:  restic
Namespaces:
  Included:  *
  Excluded:  <none>
Resources:
  Included:        *
  Excluded:        nodes, events, events.events.k8s.io
  Cluster-scoped:  auto
Namespace mappings:  <none>
Label selector:  <none>
Restore PVs:  auto
Phase:  Completed
Validation errors:  <none>

Warnings:  <error getting warnings: Get http://minio.heptio-ark.svc:9000/ark/restic/restore-restic-20180627115342-results.gz?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Credential=minio%2F20180627%2Fminio%2Fs3%2Faws4_request&X-Amz-Date=20180627T095449Z&X-Amz-Expires=600&X-Amz-SignedHeaders=host&X-Amz-Signature=08f84581b2ef7f32a611b2e67779dc9692ccf9c965b10422f72640f28cd6d49e: dial tcp: lookup minio.heptio-ark.svc on 8.8.8.8:53: no such host>

Errors:  <error getting errors: Get http://minio.heptio-ark.svc:9000/ark/restic/restore-restic-20180627115342-results.gz?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Credential=minio%2F20180627%2Fminio%2Fs3%2Faws4_request&X-Amz-Date=20180627T095449Z&X-Amz-Expires=600&X-Amz-SignedHeaders=host&X-Amz-Signature=08f84581b2ef7f32a611b2e67779dc9692ccf9c965b10422f72640f28cd6d49e: dial tcp: lookup minio.heptio-ark.svc on 8.8.8.8:53: no such host>
```

Do you know why this error arises?

**Comment by ncdc** [comment]
See also: #387.

**Comment by ncdc** [comment]
@domcar if you run `ark restore get`, what are the warning and error counts? Can you exec into pods restored with restic data and see if the data were restored? I'm wondering if there was an error with the restore, or is it possible that it succeeded and the problem is that describe isn't working?

**Comment by domcar** [comment]
I also spun another cluster at GKE (but using Minio for the bucket), and exactly the same problem with the minio service appeared `dial tcp: lookup minio.heptio-ark.svc on 8.8.8.8:53: no such host>`

**Comment by ncdc** [comment]
@domcar if you're using Minio from our example, you end up with a `ClusterIP` service for it, sitting in front of a pod. The service (`minio.heptio-ark.svc`) is not routable from outside the cluster. What we really should do is add a second, optional URL to the AWS `backupStorageProvider` configuration, something like `publicUrl`, that Ark can use when returning the pre-signed URLs to download the backup tarball, backup logs, restore logs, and restore warnings/errors (part of `ark restore describe`).

Could you please `kubectl describe` the `PersistentVolume` that's part of the restore attempt? Thanks!

**Comment by domcar** [comment]
@ncdc yes `minio.heptio-ark.svc` is not routable from outside but still the restore process should get to it, right? The process is inside the cluster, that's why I don't understand the `dial tcp: lookup minio.heptio-ark.svc on 8.8.8.8:53: no such host`.

Anyway, i'm trying with `NodePort` now. My setup is very simple, 1storage class, 1 PVC, 1 deployment, there is nothing else on the cluster.
 
This is the output of `kubectl describe pv `
```
Name:            pvc-5b022e8f-7b78-11e8-b458-0a580af40389
Labels:          <none>
Annotations:     kubernetes.io/createdby=vsphere-volume-dynamic-provisioner
                 pv.kubernetes.io/bound-by-controller=yes
                 pv.kubernetes.io/provisioned-by=kubernetes.io/vsphere-volume
StorageClass:    sphere
Status:          Bound
Claim:           default/sphere-pvc
Reclaim Policy:  Delete
Access Modes:    RWO
Capacity:        1Gi
Message:
Source:
    Type:        vSphereVolume (a Persistent Disk resource in vSphere)
    VolumePath:      StoragePolicyName:  %v

    FSType:                            [cf2-cvc02-kubermatic-poc] kubevols/kubernetes-dynamic-pvc-5b022e8f-7b78-11e8-b458-0a580af40389.vmdk
%!(EXTRA string=ext4, string=)Events:  <none>
```

This is the podvolumebackup:
```
apiVersion: v1
items:
- apiVersion: ark.heptio.com/v1
  kind: PodVolumeBackup
  metadata:
    clusterName: ""
    creationTimestamp: 2018-06-29T08:44:31Z
    generateName: restic-test-
    generation: 1
    labels:
      ark.heptio.com/backup-name: restic-test
      ark.heptio.com/backup-uid: a275e5dc-7b78-11e8-b458-0a580af40389
    name: restic-test-wm6c9
    namespace: heptio-ark
    ownerReferences:
    - apiVersion: ark.heptio.com/v1
      controller: true
      kind: Backup
      name: restic-test
      uid: a275e5dc-7b78-11e8-b458-0a580af40389
    resourceVersion: "169212"
    selfLink: /apis/ark.heptio.com/v1/namespaces/heptio-ark/podvolumebackups/restic-test-wm6c9
    uid: a4555f26-7b78-11e8-b458-0a580af40389
  spec:
    node: kubermatic-bvt7mbts7d-kk2xw
    pod:
      kind: Pod
      name: nginx-test-645796c976-6jm7n
      namespace: default
      uid: 5b1a4ec1-7b78-11e8-b458-0a580af40389
    repoPrefix: s3:http://62.80.0.53:32199/restic-bucket
    tags:
      backup: restic-test
      backup-uid: a275e5dc-7b78-11e8-b458-0a580af40389
      ns: default
      pod: nginx-test-645796c976-6jm7n
      pod-uid: 5b1a4ec1-7b78-11e8-b458-0a580af40389
      volume: persistent-storage
    volume: persistent-storage
  status:
    message: ""
    path: /host_pods/5b1a4ec1-7b78-11e8-b458-0a580af40389/volumes/kubernetes.io~vsphere-volume/pvc-5b022e8f-7b78-11e8-b458-0a580af40389
    phase: Completed
    snapshotID: 3ae0c589
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

I then delete Deployment and PVC and I try a restore `ark restore create --from-backup restic-test`

`ark restore describe restic-test-20180629105304` shows 
```
Name:         restic-test-20180629105304
Namespace:    heptio-ark
Labels:       <none>
Annotations:  <none>
Backup:  restic-test
Namespaces:
  Included:  *
  Excluded:  <none>
Resources:
  Included:        *
  Excluded:        nodes, events, events.events.k8s.io
  Cluster-scoped:  auto
Namespace mappings:  <none>
Label selector:  <none>
Restore PVs:  auto
Phase:  Completed
Validation errors:  <none>
Warnings:  <none>
Errors:    <none>
```

`PVC` and `PV` are there and bound, the Pod is stuck in `init:o/1`:

```
kubectl get pvc
NAME         STATUS    VOLUME   CAPACITY   ACCESS MODES   STORAGECLASS   AGE
sphere-pvc   Bound  pvc-5b022e8f-7b78-11e8-b458-0a580af40389   1Gi   RWO   sphere    1m
```
```
kubectl get pv
NAME CAPACITY ACCESS MODES RECLAIM POLICY STATUS CLAIM STORAGECLASS REASON AGE
pvc-5b022e8f-7b78-11e8-b458-0a580af40389   1Gi  RWO    Delete   Bound   default/sphere-pvc  1m
```
```
kubectl get pod
NAME                          READY     STATUS     RESTARTS   AGE
nginx-test-645796c976-6jm7n   0/1       Init:0/1   0          1m
```
`kubectl describe pod`

```
Events:
  Type     Reason                 Age               From                                  Message
  ----     ------                 ----              ----                                  -------
  Normal   Scheduled              5m                default-scheduler                     Successfully assigned nginx-test-645796c976-6jm7n to kubermatic-bvt7mbts7d-kk2xw
  Normal   SuccessfulMountVolume  5m                kubelet, kubermatic-bvt7mbts7d-kk2xw  MountVolume.SetUp succeeded for volume "default-token-zdksd"
  Warning  FailedAttachVolume     1m (x10 over 5m)  attachdetach-controller               AttachVolume.Attach failed for volume "pvc-5b022e8f-7b78-11e8-b458-0a580af40389" : File []/vmfs/volumes/5c24140e-8a01a62c/kubevols/kubernetes-dynamic-pvc-5b022e8f-7b78-11e8-b458-0a580af40389.vmdk was not found
  Warning  FailedMount            1m (x2 over 3m)   kubelet, kubermatic-bvt7mbts7d-kk2xw  Unable to mount volumes for pod "nginx-test-645796c976-6jm7n_default(d6deba31-7b79-11e8-b458-0a580af40389)": timeout expired waiting for volumes to attach or mount for pod "default"/"nginx-test-645796c976-6jm7n". list of unmounted volumes=[persistent-storage]. list of unattached volumes=[persistent-storage default-token-zdksd]
```

**Comment by ncdc** [comment]
> yes minio.heptio-ark.svc is not routable from outside but still the restore process should get to it, right? The process is inside the cluster, that's why I don't understand the dial tcp: lookup minio.heptio-ark.svc on 8.8.8.8:53: no such host.

The "no such host" error that you're seeing is from the computer where you are running `ark restore describe` - that command asks the ark server for a URL it can use to download the file from object storage that contains the details about the warnings and the errors for the restore. The error that you're seeing:

```
Warnings:  <error getting warnings: Get http://minio.heptio-ark.svc:9000/ark/restic/restore-restic-20180627115342-results.gz?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Credential=minio%2F20180627%2Fminio%2Fs3%2Faws4_request&X-Amz-Date=20180627T095449Z&X-Amz-Expires=600&X-Amz-SignedHeaders=host&X-Amz-Signature=08f84581b2ef7f32a611b2e67779dc9692ccf9c965b10422f72640f28cd6d49e: dial tcp: lookup minio.heptio-ark.svc on 8.8.8.8:53: no such host>
```

is only about the attempt to retrieve the warnings & errors file. The restore process itself was able to talk to minio.

I think you might be running into #613 - to confirm, could you please:

1. provide the yaml for the PV before you take the backup
1. take a backup
1. delete your PVC and deployment
1. do a restore
1. provide the yaml for the PV after it's been restored

Thanks for your patience as we work through these issues!

**Comment by domcar** [comment]
@ncdc Thanks for the explanation. 
This is the PV before snapshost:
```
apiVersion: v1
kind: PersistentVolume
metadata:
  annotations:
    kubernetes.io/createdby: vsphere-volume-dynamic-provisioner
    pv.kubernetes.io/bound-by-controller: "yes"
    pv.kubernetes.io/provisioned-by: kubernetes.io/vsphere-volume
  creationTimestamp: 2018-07-02T06:44:30Z
  finalizers:
  - kubernetes.io/pv-protection
  name: pvc-5ed64675-7dc3-11e8-b458-0a580af40389
  resourceVersion: "610906"
  selfLink: /api/v1/persistentvolumes/pvc-5ed64675-7dc3-11e8-b458-0a580af40389
  uid: 5f27212c-7dc3-11e8-b458-0a580af40389
spec:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 1Gi
  claimRef:
    apiVersion: v1
    kind: PersistentVolumeClaim
    name: sphere-pvc
    namespace: default
    resourceVersion: "610888"
    uid: 5ed64675-7dc3-11e8-b458-0a580af40389
  persistentVolumeReclaimPolicy: Delete
  storageClassName: sphere
  vsphereVolume:
    fsType: ext4
    volumePath: '[cf2-cvc02-kubermatic-poc] kubevols/kubernetes-dynamic-pvc-5ed64675-7dc3-11e8-b458-0a580af40389.vmdk'
status:
  phase: Bound
```
And this is after the restore:
```
apiVersion: v1
kind: PersistentVolume
metadata:
  annotations:
    kubernetes.io/createdby: vsphere-volume-dynamic-provisioner
    pv.kubernetes.io/bound-by-controller: "yes"
    pv.kubernetes.io/provisioned-by: kubernetes.io/vsphere-volume
  creationTimestamp: 2018-07-02T06:47:38Z
  finalizers:
  - kubernetes.io/pv-protection
  labels:
    ark-restore: restic-test-20180702084737
  name: pvc-5ed64675-7dc3-11e8-b458-0a580af40389
  resourceVersion: "611291"
  selfLink: /api/v1/persistentvolumes/pvc-5ed64675-7dc3-11e8-b458-0a580af40389
  uid: cf0a782a-7dc3-11e8-b458-0a580af40389
spec:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 1Gi
  claimRef:
    apiVersion: v1
    kind: PersistentVolumeClaim
    name: sphere-pvc
    namespace: default
    resourceVersion: "611288"
    uid: cf0fff55-7dc3-11e8-b458-0a580af40389
  persistentVolumeReclaimPolicy: Delete
  vsphereVolume:
    fsType: ext4
    volumePath: '[cf2-cvc02-kubermatic-poc] kubevols/kubernetes-dynamic-pvc-5ed64675-7dc3-11e8-b458-0a580af40389.vmdk'
status:
  phase: Bound
```

I don't know if it can be useful, these are the `describe pod nginx`
```
Name:           nginx-test-645796c976-h6rmx
Namespace:      default
Node:           kubermatic-bvt7mbts7d-kk2xw/62.80.0.53
Start Time:     Mon, 02 Jul 2018 08:47:38 +0200
Labels:         app=nginx-test
                ark-restore=restic-test-20180702084737
                pod-template-hash=2013527532
Annotations:    backup.ark.heptio.com/backup-volumes=persistent-storage
                snapshot.ark.heptio.com/persistent-storage=4459f730
Status:         Pending
IP:
Controlled By:  ReplicaSet/nginx-test-645796c976
Init Containers:
  restic-wait:
    Container ID:
    Image:         gcr.io/heptio-images/ark-restic-restore-helper:v0.9.0-alpha.2
    Image ID:
    Port:          <none>
    Args:
      ce7d56fd-7dc3-11e8-b458-0a580af40389
    State:          Waiting
      Reason:       PodInitializing
    Ready:          False
    Restart Count:  0
    Environment:
      POD_NAMESPACE:  default (v1:metadata.namespace)
      POD_NAME:       nginx-test-645796c976-h6rmx (v1:metadata.name)
    Mounts:
      /restores/persistent-storage from persistent-storage (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-zdksd (ro)
Containers:
  nginx-test:
    Container ID:
    Image:          nginx
    Image ID:
    Port:           80/TCP
    State:          Waiting
      Reason:       PodInitializing
    Ready:          False
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /usr/share/nginx/html from persistent-storage (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-zdksd (ro)
Conditions:
  Type           Status
  Initialized    False
  Ready          False
  PodScheduled   True
Volumes:
  persistent-storage:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  sphere-pvc
    ReadOnly:   false
  default-token-zdksd:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-zdksd
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     <none>
Events:
  Type     Reason                 Age               From                                  Message
  ----     ------                 ----              ----                                  -------
  Warning  FailedScheduling       2m                default-scheduler                     persistentvolumeclaim "sphere-pvc" not found
  Normal   Scheduled              2m                default-scheduler                     Successfully assigned nginx-test-645796c976-h6rmx to kubermatic-bvt7mbts7d-kk2xw
  Normal   SuccessfulMountVolume  2m                kubelet, kubermatic-bvt7mbts7d-kk2xw  MountVolume.SetUp succeeded for volume "default-token-zdksd"
  Warning  FailedMount            30s               kubelet, kubermatic-bvt7mbts7d-kk2xw  Unable to mount volumes for pod "nginx-test-645796c976-h6rmx_default(cf152e49-7dc3-11e8-b458-0a580af40389)": timeout expired waiting for volumes to attach or mount for pod "default"/"nginx-test-645796c976-h6rmx". list of unmounted volumes=[persistent-storage]. list of unattached volumes=[persistent-storage default-token-zdksd]
  Warning  FailedAttachVolume     21s (x9 over 2m)  attachdetach-controller               AttachVolume.Attach failed for volume "pvc-5ed64675-7dc3-11e8-b458-0a580af40389" : File []/vmfs/volumes/5c24140e-8a01a62c/kubevols/kubernetes-dynamic-pvc-5ed64675-7dc3-11e8-b458-0a580af40389.vmdk was not found
```

And this is the `get pod -o yaml`
```
apiVersion: v1
kind: Pod
metadata:
  annotations:
    backup.ark.heptio.com/backup-volumes: persistent-storage
    snapshot.ark.heptio.com/persistent-storage: 4459f730
  creationTimestamp: 2018-07-02T06:47:38Z
  labels:
    app: nginx-test
    ark-restore: restic-test-20180702084737
    pod-template-hash: "2013527532"
  name: nginx-test-645796c976-h6rmx
  namespace: default
  ownerReferences:
  - apiVersion: extensions/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: ReplicaSet
    name: nginx-test-645796c976
    uid: cf195c75-7dc3-11e8-b458-0a580af40389
  resourceVersion: "611306"
  selfLink: /api/v1/namespaces/default/pods/nginx-test-645796c976-h6rmx
  uid: cf152e49-7dc3-11e8-b458-0a580af40389
spec:
  containers:
  - image: nginx
    imagePullPolicy: Always
    name: nginx-test
    ports:
    - containerPort: 80
      name: http-server
      protocol: TCP
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /usr/share/nginx/html
      name: persistent-storage
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: default-token-zdksd
      readOnly: true
  dnsPolicy: ClusterFirst
  initContainers:
  - args:
    - ce7d56fd-7dc3-11e8-b458-0a580af40389
    env:
    - name: POD_NAMESPACE
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.namespace
    - name: POD_NAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.name
    image: gcr.io/heptio-images/ark-restic-restore-helper:v0.9.0-alpha.2
    imagePullPolicy: IfNotPresent
    name: restic-wait
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /restores/persistent-storage
      name: persistent-storage
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: default-token-zdksd
      readOnly: true
  nodeName: kubermatic-bvt7mbts7d-kk2xw
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 30
  volumes:
  - name: persistent-storage
    persistentVolumeClaim:
      claimName: sphere-pvc
  - name: default-token-zdksd
    secret:
      defaultMode: 420
      secretName: default-token-zdksd
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: 2018-07-02T06:47:38Z
    message: 'containers with incomplete status: [restic-wait]'
    reason: ContainersNotInitialized
    status: "False"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: 2018-07-02T06:47:38Z
    message: 'containers with unready status: [nginx-test]'
    reason: ContainersNotReady
    status: "False"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: 2018-07-02T06:47:38Z
    status: "True"
    type: PodScheduled
  containerStatuses:
  - image: nginx
    imageID: ""
    lastState: {}
    name: nginx-test
    ready: false
    restartCount: 0
    state:
      waiting:
        reason: PodInitializing
  hostIP: 62.80.0.53
  initContainerStatuses:
  - image: gcr.io/heptio-images/ark-restic-restore-helper:v0.9.0-alpha.2
    imageID: ""
    lastState: {}
    name: restic-wait
    ready: false
    restartCount: 0
    state:
      waiting:
        reason: PodInitializing
  phase: Pending
  qosClass: BestEffort
  startTime: 2018-07-02T06:47:38Z
```

**Comment by ncdc** [comment]
@domcar ok, yes, this appears to be fixed by #613. Please try v0.9.0-alpha.3 and let us know if it resolves the issue. Note, some of the internals have changed - see https://github.com/heptio/ark/blob/master/docs/restic.md#pre-install-steps---upgrading-from-a-previous-v09-alpha for more details on what to do.

**Comment by domcar** [comment]
@ncdc Thanks a lot, that was the problem. I tried now with the `v0.9.0-alpha.3` and it worked.
I'll close this Issue.
